### DISCUSSION TITLE
Any updates on single frequency analysis?

### DISCUSSION THREAD

**Discussion by tomhajjar** [ra3xdh/qucs_s#761] [Q&A] [answered]
Has anything changed trying to plot a single frequency, CW mode? It was discussed in #156, #161 and #386. The solution was to use the 3D plot or Nutmeg script. Trying to manipulate XYZ axis to get a readable plot isn't easy and making a Nutmeg script is tedious and error prone especially for me. The Nutmeg script required to do the attached analysis requires10x the time.

![2024-06-16_204930](https://github.com/ra3xdh/qucs_s/assets/102838604/68b612d9-f75d-4413-a2c5-1af123a3df60)

[BB814_CV_curve_S11.zip](https://github.com/user-attachments/files/15860494/BB814_CV_curve_S11.zip)

**Comment by ra3xdh** [comment, marked as answer]
You don't need to use 3D plots to simulate the CV-curve. I have uploaded an example using script simulation type and vector assembling  See the file `share/qucs-s/examples/ngspice/CV_curve.sch` See also this comment: https://github.com/ra3xdh/qucs_s/issues/277#issuecomment-1575473862 Also see #277 for solution. Unfortunately I have no idea how to automate this and get rid of Nutmeg script.

**Reply by tomhajjar** [reply]
I have been using CV_curve.sch for a while but I couldn't get this technique to work for the dual varactor sub-circuit. Using the method using S-parameters worked first try.

![2024-06-17_093957](https://github.com/ra3xdh/qucs_s/assets/102838604/02cb2dd9-8f12-4fc9-b16d-a30383c797a0)

[BB814_CV_curve_mod.zip](https://github.com/user-attachments/files/15871940/BB814_CV_curve_mod.zip)

**Comment by dwarning** [comment]
Life would be much easier if we can use OP variables which are provided by most models in ngspice. The attached netlist show us the CV curve by a simple dc sweep.

[CV_curve_by_OP_variables.txt](https://github.com/user-attachments/files/15871191/CV_curve_by_OP_variables.txt)

![Bildschirmfoto vom 2024-06-17 14-50-43](https://github.com/ra3xdh/qucs_s/assets/18441932/7981bcce-e2a9-41ec-9161-6c0e53ff9949)

But qucs-s schematic as follows:
![Bildschirmfoto vom 2024-06-17 15-11-01](https://github.com/ra3xdh/qucs_s/assets/18441932/a7b354f8-15d0-4bb7-bbda-542712c4f1ea)

show no variables by using qucs-s dc and parameter sweep command. With the shown nutmeg script the curve comes very well.

This feature would be really promising for plotting Vth, gm, gds, Beta, charges and much more in a simple way and has the ability to make parameter sweeps over operating points and temperature.

**Reply by tomhajjar** [reply]
In my case, I also wanted to know the effects of packaging parasitics at the operating frequency. The CV curve shows if the parasitics are causing issues. It gets bumps of turns inductive at a resonance. Here is the BB639 using the manufacturers die and package Spice data.

[BB639_Varactor_ngspice_prj.zip](https://github.com/user-attachments/files/15872442/BB639_Varactor_ngspice_prj.zip)

**Comment by ra3xdh** [comment]
@dwarning , I will think how to dump the OP variables from each device automatically. This may be technically possible. This will require a netlister extension. I cannot promise a quick result here. It may be needed to introduce a boolean type simulation option like "Dump OP vars". The default should be false, because the variable list will be overcrowded. 

@tomhajjar 

> I have been using CV_curve.sch for a while but I couldn't get this technique to work for the dual varactor sub-circuit

You have an error in the script. You have forgot to change the AC source name. The current should be `V2#bracnh`. Original example had two sources in series and there was the same current flowing through AC and DC source.  Everything works as expected after applying this correction.

![image](https://github.com/ra3xdh/qucs_s/assets/4920080/9de9bf72-cc37-4fda-b68d-4d489c9fff93)

**Comment by dwarning** [comment]
@ra3xdh 
Perhaps you create a branch dump_op_vars. So I can build and test it before merge into master branch.

**Reply by ra3xdh** [reply]
@dwarning I have created a task #789. The OP variables dump is planned for one of the next releases.
